### PR TITLE
Issue 1461

### PR DESCRIPTION
**Pull request by 4r9h**
@rafrombrc @trink could you conduct initial review?

### CODE REVIEW

File: sandbox/lua/modules/graphite.lua [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+--[[
+Module contains utilities for Graphite
+--]]
+-- Imports
+require "string"
+require "math"
+require "table"
+
+local M = {}
+
```
Comment (trink) [inline sandbox/lua/modules/graphite.lua]: I knew I should have updated alert/annotation to use environments (see anomaly.lua).  https://github.com/mozilla-services/heka/issues/1496

It is too easy to leak into the global environment if you aren't careful (like thresh_bound and num_in_thresh below).

Comment (4r9h) [inline sandbox/lua/modules/graphite.lua]: Yeah, i'm relatively new to lua and alert.lua was my entrypoint. I've spotted 3 different conventions for modules and i wasn't sure which standard here. ;-) Can i start working on #1496 after i'll clean up this PR?

Comment (trink) [inline sandbox/lua/modules/graphite.lua]: Already got it, I'll put a PR in.

Comment (trink) [inline sandbox/lua/modules/graphite.lua]: I was playing with some code to address 1496 but the original hack of not using environments will have to stand for now.

Comment (4r9h) [inline sandbox/lua/modules/graphite.lua]: Okay, then i'll update  PR (with docstrings)  later

On Thu, Apr 23, 2015 at 4:10 PM, Mike Trinkala <notifications@github.com>
wrote:

> In sandbox/lua/modules/graphite.lua
> <https://github.com/mozilla-services/heka/pull/1492#discussion_r28965410>:
>
> > @@ -0,0 +1,135 @@
> > +-- This Source Code Form is subject to the terms of the Mozilla Public
> > +-- License, v. 2.0. If a copy of the MPL was not distributed with this
> > +-- file, You can obtain one at http://mozilla.org/MPL/2.0/.
> > +
> > +--[[
> > +Module contains utilities for Graphite
> > +--]]
> > +-- Imports
> > +require "string"
> > +require "math"
> > +require "table"
> > +
> > +local M = {}
> > +
>
> I was playing with some code to address 1496 but the original hack of not
> using environments will have to stand for now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mozilla-services/heka/pull/1492/files#r28965410>.
>



-- 
jarek@reijutsu:~$ fortune
You have Egyptian flu: you're going to be a mummy.
jarek@reijutsu:~$ fortune
You now have Asian Flu.

Comment (trink) [inline sandbox/lua/modules/graphite.lua]: This plugin doesn't use the global preservation hack and can be written using the preferred environment method (I was just clarifying the other alert and anomaly must remain the same).

Comment (4r9h) [inline sandbox/lua/modules/graphite.lua]: Could you check again?

**Comment by rafrombrc** [comment]
Overall this looks good. All that's missing is a changelog entry (easy) and integration with the existing documentation (still easy, but a bit more work).

For documentation, you'll want to take a look at the documentation comments for the other modules, and how they're all woven together into the sandbox docs in the `modules.rst` file.

Thanks!

**Comment by 4r9h** [comment]
@rafrombrc I've updated PR with docs. Any language fixes/improvements are welcome (as always).

File: sandbox/lua/modules/graphite.lua [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+						bucket, sum, now_sec,
+						bucket, mean, now_sec,
+						bucket, percent_thresh, mean_percentile, now_sec,
+						bucket, percent_thresh, upper_percentile, now_sec)
+end
+
+function multi_timeseries_metrics(timers, ticker_interval, percent_thresh, now_sec)
+	local stats = {}
+    for bucket, times in pairs(timers) do
+        stats[#stats+1] = M.timeseries_metrics(bucket, times, ticker_interval, percent_thresh, now_sec)
```
Comment (trink) [inline sandbox/lua/modules/graphite.lua]: Not critical but a couple of M's can still be removed (here and line 101)

File: sandbox/lua/testsupport/graphite.lua [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    add_to_payload(graphite.multi_counts_rates(status_codes, ticker_interval, now_sec))
+    add_to_payload(graphite.multi_timeseries_metrics(request_times, ticker_interval, percent_thresh, now_sec))
+
+    for bucket, times in pairs(request_times) do
+        num_stats = num_stats + 1
+    end
+
+    add_to_payload(string.format("stats.statsd.numStats %d %d\n", num_stats, now_sec))
+    inject_payload("txt", "statmetric")
+end
```
Comment (trink) [inline sandbox/lua/testsupport/graphite.lua]: Also not critical, add an new line to the end of the files.

Comment (4r9h) [inline sandbox/lua/testsupport/graphite.lua]: Sure, will do.

On Mon, Apr 27, 2015 at 3:53 PM, Mike Trinkala <notifications@github.com>
wrote:

> In sandbox/lua/testsupport/graphite.lua
> <https://github.com/mozilla-services/heka/pull/1492#discussion_r29147392>:
>
> > +    local num_stats = 0
> > +
> > +    for bucket, count in pairs(status_codes) do
> > +        num_stats = num_stats + 1
> > +    end
> > +
> > +    add_to_payload(graphite.multi_counts_rates(status_codes, ticker_interval, now_sec))
> > +    add_to_payload(graphite.multi_timeseries_metrics(request_times, ticker_interval, percent_thresh, now_sec))
> > +
> > +    for bucket, times in pairs(request_times) do
> > +        num_stats = num_stats + 1
> > +    end
> > +
> > +    add_to_payload(string.format("stats.statsd.numStats %d %d\n", num_stats, now_sec))
> > +    inject_payload("txt", "statmetric")
> > +end
>
> Also not critical, add an new line to the end of the files.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mozilla-services/heka/pull/1492/files#r29147392>.
>



-- 
jarek@reijutsu:~$ fortune
You have Egyptian flu: you're going to be a mummy.
jarek@reijutsu:~$ fortune
You now have Asian Flu.

**Comment by 4r9h** [comment]
@trink ok, updated.

**Comment by trink** [comment]
looks good r+

**Pull request merged**